pkg/storage/mvcc.go

```diff
@@ -4301,6 +4361,305 @@ func MVCCGarbageCollect(
+			UpperBound: gcKey.LatchSpan.EndKey,
+			KeyTypes:   IterKeyTypeRangesOnly,
+		})
+
+		iter.SeekGE(MVCCKey{Key: gcKey.LatchSpan.Key})
+
+		for ; ; iter.Next() {
+			if ok, err := iter.Valid(); err != nil {
+				return err
+			} else if !ok {
+				break
+			}
+
+			bounds := iter.RangeBounds()
+			unsafeRangeKeys := iter.RangeKeys()
+
+			// Check if preceeding range tombstone is adjacent to GC'd one. If we
+			// started iterating too early, just skip to next key. If boundaries
+			// match, then we capture start of the range and timestamps for later.
+			switch bounds.EndKey.Compare(gcKey.StartKey) {
+			case -1:
+				continue
+			case 0:
+				mergeTracker.update(bounds.Key, bounds.EndKey, unsafeRangeKeys)
+				continue
+			}
+
+			// Terminate loop once we've reached a range tombstone past the right
+			// GC range key boundary.
+			if cmp := bounds.Key.Compare(gcKey.EndKey); cmp >= 0 {
+				mergeTracker.update(bounds.Key, bounds.EndKey, unsafeRangeKeys)
+				break
+			}
+
+			// Check if we have a partial overlap between range tombstone and
+			// requested GCd range. This shouldn't happen in most cases, but we can
+			// have a range merge between GC run and cmd_gc execution or
+			// erroneous GC request and mvcc stats should be updated correctly.
+			// In those cases gcKey boundaries will not match underlying range
+			// tombstone boundaries and need to be adjusted.
+			trimLeft, trimRight := false, false
+			candidateStartKey := bounds.Key
+			if gcKey.StartKey.Compare(candidateStartKey) > 0 {
+				candidateStartKey = gcKey.StartKey
+				trimLeft = true
+			}
+			candidateEndKey := bounds.EndKey
+			if gcKey.EndKey.Compare(candidateEndKey) < 0 {
+				candidateEndKey = gcKey.EndKey
+				trimRight = true
+			}
+
+			removedRange := MVCCRangeKeyValue{
+				RangeKey: MVCCRangeKey{
+					StartKey: candidateStartKey,
+					EndKey:   candidateEndKey,
+				},
+			}
+			remaining := len(unsafeRangeKeys)
+			for i, rkv := range unsafeRangeKeys {
+				removedRange.RangeKey.Timestamp = rkv.RangeKey.Timestamp
+				removedRange.Value = rkv.Value
+				remove := rkv.RangeKey.Timestamp.LessEq(gcKey.Timestamp)
+				if remove {
+					if err := rw.ClearMVCCRangeKey(removedRange.RangeKey); err != nil {
+						return err
+					}
+					remaining--
+				}
+				if ms != nil {
+					ms.Add(updateStatsOnRangeTombstoneGC(removedRange, trimLeft, trimRight, i == 0, remove))
+				}
+			}
+
+			mergeTracker.update(candidateStartKey, candidateEndKey, unsafeRangeKeys[0:remaining])
+
+			// Verify that there are no remaining data under the deleted range using
+			// time bound iterator.
+			ptIter = NewMVCCIncrementalIterator(rw, MVCCIncrementalIterOptions{
+				KeyTypes:     IterKeyTypePointsOnly,
+				StartKey:     gcKey.StartKey,
+				EndKey:       gcKey.EndKey,
+				EndTime:      gcKey.Timestamp,
+				IntentPolicy: MVCCIncrementalIterIntentPolicyEmit,
+			})
+			ptIter.SeekGE(MVCCKey{Key: candidateStartKey})
+			for ; ; ptIter.Next() {
```

Comment:
Do we need all these checks? Wouldn't it be sufficient to check `ptIter.Valid()` after the seek, and if it's `true` then just return an error with the key? We expect this span to be completely empty, right?

Reply:
If we have a tombstone under the range tombstone it should be fine as there would be no values exposed. Code also checks for meta timestamp if it has fallen under the range tombstone. That should never happen though if delete using range tombstone has no bugs.
I don't think it is unreasonable to check as in most cases we'll just get !ok after seek and we are done.

Reply:
Yeah sure, that's fine. The intent check seems reasonable, although I think we'd hit the provisional value in any case (which I think can only move forwards in time when it gets resolved).

But in that case, shouldn't we accept all keys that are hidden by tombstones too? Like, if we have a tombstone at `a@5` but a real key at `a@4` then this code is going to error anyway, so it seems like we should allow that scenario then.

But we'll need to make changes here when we implement the `ClearRange` optimization anyway, so this is fine for now I suppose.

Reply:
Yeah, we expect all point keys below here to be removed by point GC anyway, and if we were to reveal any point tombstones then it wouldn't be a correctness problem, so it's fine.